src/hazelcore/extractions/ocaml/OCamlExtraction_Exp.re

```diff
@@ -0,0 +1,270 @@
+    | Var(x) =>
+      let ocaml_dp = OCamlExtraction_Pat.extract(~dp);
+      let ocaml_de1 = extract(~ctx, ~de=de1);
+      let ctx' = Contexts.extend_gamma(ctx, (x, snd(ocaml_de1)));
+      let ocaml_de2 = extract(~ctx=ctx', ~de=de2);
+      // use the "let .. = ..;;" format
+      // if it's a recursive function, it's format of let ... = fix ..., we change it to "let rec"
+      //, because the recursive function should be declared by "rec" only in ocaml
+      switch (de1) {
+      | FixF(_, _, _) => (
+          "let rec "
+          ++ ocaml_dp
+          ++ " : "
+          ++ OCamlExtraction_Typ.extract(~t=snd(ocaml_de1))
+          ++ " = "
+          ++ fst(ocaml_de1)
+          ++ ";;\n"
+          ++ fst(ocaml_de2),
+          snd(ocaml_de2),
+          // fst(ocaml_de1) ++ fst(ocaml_de2),
+          // snd(ocaml_de2),
+        )
+      | _ => (
+          "let "
+          ++ ocaml_dp
+          ++ " : "
+          ++ OCamlExtraction_Typ.extract(~t=snd(ocaml_de1))
+          ++ " = "
+          ++ fst(ocaml_de1)
+          ++ ";;\n"
+          ++ fst(ocaml_de2),
+          snd(ocaml_de2),
+        )
+      };
+    | _ => failwith("Exp: Let, not a variable pattern")
+    }
+  | FixF(x, ht, de) =>
+    //it's the case of fixpoint/ recursive functions
+    //Hazel programs will only ever result in internal expressions containing let ... = fix ...
+    //let rec f: some type = fun x -> f ... (is reasonable)
+    // x should be already inserted into context
+
+    //let ocaml_ht = OCamlExtraction_Typ.extract(~t=ht);
+    let ctx' = Contexts.extend_gamma(ctx, (x, ht));
+    let ocaml_de = extract(~ctx=ctx', ~de);
+    (
+      // "let rec " ++ x ++ " : " ++ ocaml_ht ++ " = " ++ fst(ocaml_de) ++ ";;\n",
+      fst(ocaml_de),
+      ht,
+    );
+
+  | Lam(dp, ht, de) =>
+    //Htyp.t is the ground type of pattern
+    switch (dp) {
+    | Var(x) =>
+      let ocaml_dp = OCamlExtraction_Pat.extract(~dp);
+      let ocaml_ht = OCamlExtraction_Typ.extract(~t=ht);
+      let ctx' = Contexts.extend_gamma(ctx, (x, ht));
+      let ocaml_de = extract(~ctx=ctx', ~de);
+      (
+        "(fun ("
+        ++ ocaml_dp
+        ++ " : "
+        ++ ocaml_ht
+        ++ ") -> ("
+        ++ fst(ocaml_de)
+        ++ "))",
+        Arrow(ht, snd(ocaml_de)),
+      );
+    | _ => failwith("Exp: Lambda not variable")
+    }
+  | Ap(de1, de2) =>
+    //apply
+    let ocaml_de1 = extract(~ctx, ~de=de1);
+    let ocaml_de2 = extract(~ctx, ~de=de2);
+    switch (snd(ocaml_de1)) {
+    | Arrow(t1, t2) =>
+      // Hole -> some type should be ok
+      if (t1 == snd(ocaml_de2) || t1 == Hole) {
+        ("(" ++ fst(ocaml_de1) ++ " " ++ fst(ocaml_de2) ++ ")", t2);
+      } else {
+        failwith("Exp: Apply, type inconsistent");
+      }
+    | _ => failwith("Exp: Apply, not an Arrow type")
+    };
+  | BoolLit(bool) => (string_of_bool(bool), Bool)
+  | IntLit(int) => (string_of_int(int), Int)
+  | FloatLit(float) => (string_of_float(float), Float)
```

Comment:
will need to special case special floats like nan and inf (ask Yunsoo for help if you need it)